test/framework/template.go

```diff
@@ -38,12 +38,20 @@ type templateRegistry struct {
 // 3. If the template doesn't exist, default to the value of the default template env vars: eg. "T_CLOUDSTACK_TEMPLATE_REDHAT_1_23".
 // This is a catch all condition. Mostly for edge cases where the bundle has been updated with a new eks-d version, but the
 // the new template hasn't been imported yet. It also preserves backwards compatibility.
-func (tc *templateRegistry) templateForRelease(t *testing.T, osFamily anywherev1.OSFamily, release *releasev1.EksARelease, kubeVersion anywherev1.KubernetesVersion) string {
+func (tc *templateRegistry) templateForRelease(t *testing.T, osFamily anywherev1.OSFamily, release *releasev1.EksARelease, kubeVersion anywherev1.KubernetesVersion, osVersion ...string) string {
 	t.Helper()
 	osFamilyStr := string(osFamily)
 	versionsBundle := readVersionsBundles(t, release, kubeVersion)
 	eksDName := versionsBundle.EksD.Name
-	templateEnvVarName := tc.generator.envVarForTemplate(osFamilyStr, eksDName)
+
+	// If an OS version is not passed in, then we use the default version for this OS family.
+	// Otherwise we use the OS version passed in
+	osVersionStr := defaultOSVersions[osFamilyStr]
```

Comment:
Is it possible to be explicit and pass the OS version always? Instead of having to use the variadic arguments and having this defaulting here. It might seems simpler to use, but it's far from obvious to a new reader and this logic is quite burried. Plus it's not documented anywhere.

Either that or just have one argument, like `OS`. And then be able to derive the OSFamily from the OS (which each provider can easily do).

Reply:
Parametrizing the OS version in tests is a new system so I felt it would be better to also support the existing way of writing tests that is less strict and can handle defaults. So the main reason is backwards compatibility. I can document it in the E2E readme if that's okay.

Reply:
ah yeah I get that, this is new. What I'm trying to say is I think it's better to make it required. I propose simplicity + explicit vs "easy" to use.

Reply:
I don't believe that making it variadic makes it more complex, it's just allowing us to do the defaulting and for the most part is abstracted from the user. I can change it though, but can I take it as a followup? Otherwise it would require touching every method out there.

Reply:
I also added documentation in the e2e readme for configuring the OS version, so I feel that should help a new reader until we make it explicit.

Reply:
OK, what about something like this.

Instead of considering osFamily and osVersion two different things, we just make them one. They would only make sense as two arguments if the versions could be mix and matched in between os families, but they can't, they are exclusive for each type of OS.

So I believe we can start thinking about an OS as the combination of its family and its version.

```go
type OS string

const (
	Ubuntu2204    OS = "ubuntu-22-04"
	Ubuntu2004    OS = "ubuntu" // We don't add the version number in the string to facilitate backwards compatibility
	Bottlerocket1 OS = "br"     // We don't add the version number in the string to facilitate backwards compatibility
	RedHat8       OS = "redhat" // We don't add the version number in the string to facilitate backwards compatibility
)
```

Then, the signature of the methods is simplified, and the actual implementation too.

```go
// providerTemplateNameGenerator is an interface for getting template name for each provider.
type providerTemplateNameGenerator interface {
	envVarForTemplate(os OS, eksDName string) string
	defaultNameForTemplate(os OS, eksDName string) string
	searchTemplate(ctx context.Context, template string) (string, error)
	defaultEnvVarForTemplate(os OS, kubeVersion anywherev1.KubernetesVersion) string
}

func (tc *templateRegistry) templateForRelease(t *testing.T, operatingSystem OS, release *releasev1.EksARelease, kubeVersion anywherev1.KubernetesVersion) string {
```

This also simplifies the methods to update cluster specs:
```go
func (v *VSphere) WithKubeVersionAndOS(kubeVersion anywherev1.KubernetesVersion, os OS) api.ClusterConfigFiller {
	return api.JoinClusterConfigFillers(
		api.ClusterToConfigFiller(api.WithKubernetesVersion(kubeVersion)),
		api.VSphereToConfigFiller(
			v.templateForKubeVersionAndOS(os, kubeVersion),
			api.WithOsFamilyForAllMachines(osFamiliesForOS[os]),
		),
	)
}
```

Btw, in order to extract the osFamily from the OS:
```go
var osFamiliesForOS = map[OS]anywherev1.OSFamily{
	Ubuntu2204:    anywherev1.Ubuntu,
	Ubuntu2004:    anywherev1.Ubuntu,
	Bottlerocket1: anywherev1.Bottlerocket,
	RedHat8:       anywherev1.RedHat,
}
```
And this can be done transparently to the test user, so it's one concern less.

And finally, this is how the tests read:

```go
func TestVSphereKubernetes127Ubuntu2204SimpleFlow(t *testing.T) {
	vsphere := framework.NewVSphere(t)
	test := framework.NewClusterE2ETest(t, vsphere)
	test.WithClusterConfig(
		vsphere.WithKubeVersionAndOS(v1alpha1.Kube127, framework.Ubuntu2204),
	)

	runSimpleFlow(test)
}
```

Which allows to parametrize tests even more (not asking to make this change, but just an example of how clean it can read):
```go
func TestVSphereKubernetes127Ubuntu2204SimpleFlow(t *testing.T) {
	runTestWithSimpleFlow(
		t,
		framework.NewVSphere(t),
		v1alpha1.Kube127,
		framework.Ubuntu2204,
	)
}

func runTestWithSimpleFlow(t *testing.T, provider framework.Provider, kubeVersion v1alpha1.KubernetesVersion, os framework.OS) {
	test := framework.NewClusterE2ETest(t, provider)
	test.WithClusterConfig(
		provider.WithKubeVersionAndOS(kubeVersion, os),
	)

	runSimpleFlow(test)
}
```

Reply:
partial implementation, in case it's easier to read: https://github.com/g-gaston/eks-anywhere/pull/new/ubuntu-22.04

Reply:
Thanks for the guidance with examples, super clean! I see that more recent tests are using the above patter for provider instantiation without the provider options, while the older ones are using the options.
```
vsphere := framework.NewVSphere(t)
```
Did you want me to remove the provider options for kubeVersionAndOS in the scope of this PR?

Reply:
> Did you want me to remove the provider options for kubeVersionAndOS

Ah nono, that's a lot. I'm only suggesting writing the new tests this way.
All the other provider options can default to the original "version" for that OS, since that's what all the existing tests expect any way.

That way we only scope this changes to your new tests.

Reply:
Done!

Reply:
Thanks a lot! I know this was a painful change but love the result!